_posts/2019-03-30-numba-stencil.md

```diff
@@ -0,0 +1,323 @@
+---
+layout: post
+title: Composing Dask Array with Numba Stencils
+author: Matthew Rocklin
+tags: [dask, numba]
+theme: twitter
+---
+{% include JB/setup %}
+
+In this post we explore four array computing technologies, and how they
+work together to achieve powerful results.
+
+1.  Numba's stencil decorator to craft localized compute kernels
+2.  Numba's Just-In-Time (JIT) compiler for array computing in Python
+3.  Dask Array for parallelizing array computations across many chunks
+4.  NumPy's Generalized Universal Functions (gufuncs) to tie everything
+    together
+
+In the end we'll show how a novice developer can write a small amount of Python
+to efficiently compute localized computation on large amounts of data.  In
+particular we'll write a simple function to smooth images and apply that in
+parallel across a large stack of images.
+
+Here is the full code, we'll dive into it piece by piece below.
+
+```python
+import numba
+
+@numba.stencil
+def _smooth(x):
+    return (x[-1, -1] + x[-1, 0] + x[-1, 1] +
+            x[ 0, -1] + x[ 0, 0] + x[ 0, 1] +
+            x[ 1, -1] + x[ 1, 0] + x[ 1, 1]) // 9
+
+
+@numba.guvectorize(
+    [(numba.int8[:, :], numba.int8[:, :])],
+    '(n, m) -> (n, m)'
+)
+def smooth(x, out):
+    out[:] = _smooth(x)
+
+
+# If you want fake data
+import dask.array as da
+x = da.ones((1000000, 1000, 1000), chunks=('auto', -1, -1), dtype='int8')
```

Comment:
Would pick different test data that would be more interesting with a mean filter. Perhaps a checkboard pattern or something else.

Reply:
I'm inclined to keep this first post very simple and big-picture-y.  To that end I'm inclined to stick with an array that I can construct in a single line.  Constructing datasets isn't what I want to highlight, so I want to avoid much code around it.

If there is a way to construct a good dataset in a one-liner then I'm all for it.  I don't mind having this example be completely trivial though.

I think that if we end up doing a follow up that actually uses real data, that would be great.

Reply:
For the future, it would be good to think of some array examples for the `dask.datasets` module.

Reply:
Yeah, having some functions for generating test data could be good. Shipping testing data might be doable if we constrain the size a bit.

For a checkerboard pattern, normally I do something like this.

```suggestion
import numpy as np

s = (1000000, 1000, 1001)
x = da.arange(np.prod(s), chunks='auto').reshape(s) % 2
```

<br>

Mainly suggested this because it makes the change apparent to the end user if they wish to play with the code. Though this is probably more important with an example and less important with a blogpost.